### PR TITLE
add support for nested conditional rules

### PR DESCRIPTION
**Pull request by johann-sonntagbauer**
Added notes in the source code how i tackled the problem.
The main idea is to detect a nested conditional rule, extract its styles and apply them to a container conditional. 

I have added some test cases with some variations and added a test case with jss-extend

Code review and remarks are welcome. Hope to learn some new things.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -27,7 +77,7 @@ export default function jssNested({warn = consoleWarn} = {}) {
     let index
 
     for (const prop in rule.style) {
-      if (prop[0] === '&') {
+      if (isNested(prop)) {
```
Comment (kof) [inline src/index.js]: I would try to reduce the amount of function calls. You can remove all 3 functions isNEstedRule, isNesteConditional and isNested without any code duplication by simply doing this checks here and saving the result in a const:

```js
const isNested = prop[0] === '&'
const isNestedConditional = prop[0] === '@'
```

File: src/index.js [hunk 2]
```diff
@@ -19,6 +19,56 @@ export default function jssNested({warn = consoleWarn} = {}) {
...
+      .replace(refRegExp, replaceRef)
+
+    container.addRule(name, rule.style[prop], options)
+  }
+
+  function handleNestedConditional(prop, rule, container) {
+    const name = prop  // e.g. @media print
+    const containerConditionalRule = container.getRule(name)
+
+    // check if conditional rule already exists in container
```
Comment (kof) [inline src/index.js]: All my comments are a sentence.

File: src/index.js [hunk 3]
```diff
@@ -19,6 +19,56 @@ export default function jssNested({warn = consoleWarn} = {}) {
...
+    const name = prop
+      // Replace all & by the parent selector.
+      .replace(parentRegExp, rule.selector)
+      // Replace all $ref.
+      .replace(refRegExp, replaceRef)
+
+    container.addRule(name, rule.style[prop], options)
+  }
+
+  function handleNestedConditional(prop, rule, container) {
```
Comment (kof) [inline src/index.js]: I would call `rule` -> `styles`

File: src/index.js [hunk 4]
```diff
@@ -19,6 +19,56 @@ export default function jssNested({warn = consoleWarn} = {}) {
...
+    const name = prop  // e.g. @media print
```
Comment (kof) [inline src/index.js]: instead of creating a ref here, you can call it `name` in the signature

File: src/index.js (hunk 3 above)
Comment (kof) [inline src/index.js]: Maybe instead of handleNestedConditional addNestedConditional?

File: src/index.js [hunk 5]
```diff
@@ -19,6 +19,56 @@ export default function jssNested({warn = consoleWarn} = {}) {
...
+    const containerConditionalRule = container.getRule(name)
+
+    // check if conditional rule already exists in container
+    if (containerConditionalRule) {
+      // exists, so now check if we have already defined styles
+      // e.g. @media print { .some-style { color: green; } }
+      const ruleToExtend = containerConditionalRule.rules.get(rule.name)
```
Comment (kof) [inline src/index.js]: `conditionalRule.getRule(rule.name)`

File: src/index.js (hunk 3 above)
Comment (kof) [inline src/index.js]: Shouldn't we use the options as well? Because without it, the order is not preserved (index property.)

**Comment by kof** [comment]
Complete the first code review. Please let me know if you can implement the changes.

**Comment by johann-sonntagbauer** [comment]
thx for the review - I will try to add the remarks, hope to finish till sunday.

**Comment by kof** [comment]
Tuesday evening is a good day for release announcements)

**Comment by johann-sonntagbauer** [comment]
sorry for the delay I have to prepare a customer demo, so I will not find any time today. Hope to finish tomorrow. 
I have some troubles with the ordering which I have to think a little how to address them. Is the ordering of the conditionals realy important? because when the conditional is built, the ordering results of the insert order or do i miss here something fundamental. 
I want to adress this concerns with an example, but found no time to formulate. 

Thx for your support and patience.

**Comment by kof** [comment]
Wait, sorry, my bad, inside of @-rules, you are not writing full rules, it is just the props, you merge them in the right order, so it is fine.

**Comment by kof** [comment]
I can also make the changes I mentioned, its not a problem, I just didn't want to make changes in your PR without your ok.

**Comment by johann-sonntagbauer** [comment]
well I added the adaptions I did so far. Please feel free to do additional changes I am totally fine with that.

**Comment by kof** [comment]
merged

**Comment by kof** [comment]
How is it possible that you are not in githubs "contributors" list, your commits are there.

**Comment by johann-sonntagbauer** [comment]
Because of the e-mail address. I've added the address already and hope github will refresh the contributors list somehow. But never mind, it is not that big deal for me :)

**Pull request merged**